### PR TITLE
EZP-28290: Invalid cache key on url ezjscore_call_ezjscnode::subtree::2

### PR DESCRIPTION
**Pull request by andrerom**
> Issue: https://jira.ez.no/browse/EZP-28290

ping @emodric

### CODE REVIEW

**Comment by emodric** [comment]
Hmm... Not sure what is happening, but it only works 2-3 times out of ten. Other times I'm getting weird segfaults and out of memory exceptions:

```
[Sun Dec 17 21:36:38.952861 2017] [proxy_fcgi:error] [pid 17634] [client 127.0.0.1:37696] AH01071: Got error 'PHP message: PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 140476516355880 bytes) in /media/eddie/data/html/netgensite/ezpublish_legacy/kernel/classes/ezpersistentobject.php on line 855\n', referer: http://netgen.site/ngadminui/Travel
```

```
[Sun Dec 17 21:38:53.091103 2017] [proxy_fcgi:error] [pid 4335] [client 127.0.0.1:44038] AH01067: Failed to read FastCGI header, referer: http://netgen.site/ngadminui/Sports
[Sun Dec 17 21:38:53.091145 2017] [proxy_fcgi:error] [pid 4335] (104)Connection reset by peer: [client 127.0.0.1:44038] AH01075: Error dispatching request to : , referer: http://netgen.site/ngadminui/Sports
```

**Comment by emodric** [comment]
Btw, why do these URLs even go through the new stack? Shouldn't they fallback to legacy kernel?

**Comment by andrerom** [comment]
@emodric This is with legacy mode off right? In that case fallback to legacy happens after symfony routes and then url alias lookup.

**Comment by andrerom** [comment]
Any more info on your segfaults? Seems unrelated, or need some more info on that.

**Comment by emodric** [comment]
Yes, it's legacy mode off. I'm not sure if it's unrelated, those requests are the only place this is happening.

But still, shouldn't legacy router have priority over the new stack one?

**Comment by andrerom** [comment]
> But still, shouldn't legacy router have priority over the new stack one?

It's been like that since 5.0, so let's focus at the issue at hand here.

**Comment by emodric** [comment]
No problem, I was just curious :)

Anyways, the whole set of XHR requests feel sluggish in legacy in eZ Platform v2, not sure what is going on...

**Comment by emodric** [comment]
Anyway, I'm +1 for merging this PR and looking elsewhere for the cause of th OOM exceptions and segfaults, since I don't see how this specific patch could've caused them.

**Comment by andrerom** [comment]
> Anyways, the whole set of XHR requests feel sluggish in legacy in eZ Platform v2, not sure what is going on...

memcached? Sebastian reported perf issue with that. Anyway, we can look into that separately :)

**Comment by andrerom** [comment]
Review ping @alongosz 

the failures are:
- the remaining functional session failure you work on
- `each()` usage, handled in #2202

**Comment by emodric** [comment]
> memcached? Sebastian reported perf issue with that. Anyway, we can look into that separately :)

Not using memcached in dev, but I'll try to find out what is going on.

File: eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php [hunk 1]
```diff
@@ -178,7 +178,9 @@ public function removeURLAliases(array $urlAliases)
      */
     public function lookup($url)
     {
-        $cacheItem = $this->cache->getItem('ez-urlAlias-url-' . str_replace('/', '_', $url));
+        $cacheItem = $this->cache->getItem(
+            'ez-urlAlias-url-' . str_replace(['/', ':', '(', ')', '@'], ['_S', '_C', '_B', '_B', '_A'], $url)
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: Note: it's not the first time we have this issue, maybe we should consider making some common cache tags/keys sanitizer.
But for now :+1:

Comment (andrerom) [reply eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: yes, perhaps 

We can move this to abstract for places that will need it
For instance if we face the need to re introduce support for using these characters in remote ids.

**Comment by emodric** [comment]
@andrerom FYI, the issue is probably related to eZ Publish Legacy and PHP 7.2 somehow. Once I go back to 7.1, everything is much more responsive.

**Pull request merged**